Collapse x coordinates in canvas line drawer

themadcreator opened the pull request:
Adds functionality to the line drawer that will collapse very dense paths (that have a lot of line segments overlapping at the same x coordinate).

This enables canvas line rendering scale up to 100K points and beyond.

themadcreator committed:
Collapse x coordinates in canvas line drawer

hellochar commented on src/drawers/lineDrawer.ts:
```diff
@@ -75,4 +90,103 @@ export class Line extends Drawer {
+    let i = 0;
+    for (; i <= data.length; ++i) {
+      const d = data[i];
+      if (d != null) {
```
use the function returned by `d3Line.defined()` to determine whether the data point is valid or not

hellochar commented on src/drawers/lineDrawer.ts:
```diff
@@ -75,4 +90,103 @@ export class Line extends Drawer {
+
+    for (let i = iStart; i <= data.length; ++i) {
+      const d = data[i];
+      if (d == null) {
```
same comment about using defined()

hellochar commented on src/drawers/lineDrawer.ts:
```diff
@@ -75,4 +90,103 @@ export class Line extends Drawer {
+      if (d == null) {
+        continue;
+      }
+      const x = Math.floor(xFn(d, i, data));
```
this won't work if we apply a transform to the context correct?

themadcreator replied:
this will still be an effective way to bucket.

However, if we use a scale transform, the "buckets" won't abut probperly, but are we actually ever doing that?

hellochar replied:
We scale the context to fix the retina blurriness

hellochar commented on src/drawers/lineDrawer.ts:
```diff
@@ -75,4 +90,103 @@ export class Line extends Drawer {
+      const x = Math.floor(xFn(d, i, data));
+      const y = yFn(d, i, data);
+
+      if (bucket.prevX != null) {
```
can we separate the bucketing logic into a data transformation that takes in an array of data and spits out a smaller array of data, and then feed that transformed data into d3Line?

LinePlot also already does some existing data transformations - see `_filterCroppedRendering` and `_filterDownsampling` on linePlot.ts. Would be good to consolidate all the data transformation stuff into one place.

Actually wait a minute, the _filterDownsampling logic looks suspiciously like bucketing by slope? Is there some duplicated responsibility there?

themadcreator replied:
To your first point i sincerely doubt we'd get any advantage to use d3.line and we'd just causing extra GC for no reason.

themadcreator replied:
And to your other points about the stuff in lineplot, it looks like those are totally different than this method. the first just removes items that are not in the viewport (a good thing to do i think) the second will only remove data points that are on the same slope, like if you have a straight line from x,y (0, 0) -> (100, 100) with 100 datapoints, this method will draw a single line between the endpoints. it's such a minor win im curious if it's ever even used.

It's not the same logic as the stuff in the line drawer because the line may go up and down an arbitrary number of times in each bucket -- i.e. the slopes do not need to be equal.

hellochar replied:
I don't care much about d3.line, I mostly just want all our data transformation to be logically contained to one location (or to at least say why we have "data transformation in the Plot" as well as "data transformation in the Drawer")

Cool thanks for looking into the `_filterDownsampling` method. If no one's even using it maybe we can just remove/deprecate it?

hellochar reviewed (commented):
Questions about where the data transformation logic lives

themadcreator committed:
merge

themadcreator committed:
Fixing merge api refactor

Adding configuration to enable/disable dense line bucketing
Adding deferred rendering to high scale test

themadcreator committed:
Add test for coverage

hellochar commented on src/drawers/lineDrawer.ts:
```diff
@@ -56,5 +69,108 @@ export function makeLineCanvasDrawStep(d3LineFactory: () => d3.Line<any>): Canva
+  let i = 0;
+  for (; i <= data.length; ++i) {
+    const d = data[i];
+    if (d != null) {
```
use `d3Line.defined()`:
```
const xFn = d3Line.x();
const yFn = d3Line.y();
const definedFn = d3Line.defined();

...
if (definedFn(d, i, data)) {
  point
}
```

hellochar commented on src/plots/linePlot.ts:
```diff
@@ -226,8 +226,36 @@ export class Line<X> extends XYPlot<X, number> {
+   * When collapseDenseLines is enabled, vertical or nearly vertical line
+   * segments that have the same floored x coordinate will be bucketed then
+   * drawn, drastically reducing the render time of dense line plots like
+   * timeseries. Only applies to the high performance "canvas" drawer.
```
If we extract out the data-transformation logic we can use this in normal SVG lines as well

hellochar reviewed (changes requested):
Getting close

themadcreator committed:
Faster number check

themadcreator committed:
Merge branch 'develop' of github.com:palantir/plottable into bd/optimize-dense-lines

themadcreator committed:
Switch to filter method for bucketing so that SVG charts can use it too

hellochar commented on src/utils/mathUtils.ts:
```diff
@@ -71,7 +71,7 @@ export function isNaN(n: any) {
  */
 export function isValidNumber(n: any) {
-  return typeof n === "number" && !isNaN(n) && isFinite(n);
+  return typeof n === "number" && n - n < 1;
```
Is this different from before?

themadcreator replied:
same effect, much faster. it's in the inner loop of rendering so i just want to make sure we're squeeze as much perf out as possible

themadcreator committed:
Resolve conflicts
